src/optimistic_txn_db.rs

```diff
@@ -0,0 +1,360 @@
+
+use ColumnFamily;
+use DBIterator;
+use DBRawIterator;
+use DBVector;
+use Error;
+use IteratorMode;
+use Options;
+use ReadOptions;
+use WriteOptions;
+use db::Inner;
+
+use ffi;
+use libc::{c_uchar, c_int};
+use std::collections::BTreeMap;
+
+use std::ffi::CString;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::ptr;
+use transaction::Transaction;
+
+pub struct OptimisticTransactionDB {
+    pub inner: *mut ffi::rocksdb_optimistictransactiondb_t,
+    base_db: *mut ffi::rocksdb_t,
+    cfs: BTreeMap<String, ColumnFamily>,
+}
+
+impl OptimisticTransactionDB {
+    pub fn open_default<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
+        let mut options = Options::default();
+        options.create_if_missing(true);
+        Self::open(&options, path)
+    }
+
+    pub fn open<P: AsRef<Path>>(opts: &Options, path: P) -> Result<Self, Error> {
+        let path = path.as_ref();
+        let cpath = match CString::new(path.to_string_lossy().as_bytes()) {
+            Ok(c) => c,
+            Err(_) => {
+                return Err(Error::new(
+                    "Failed to convert path to CString when opening DB."
+                        .to_owned(),
+                ))
+            }
+        };
+
+        if let Err(e) = create_dir_all(&path) {
+            return Err(Error::new(
+                format!("Failed to create RocksDB directory: `{:?}`.", e),
+            ));
+        }
```

Comment:
`create_dir_all` and `cpath` creation can be moved into a separate function and used in `open` and `open_cf`.

Reply:
Removed creation directories. RocksDB does it by itself. `cpath` creation was moved into separate func.